fluent-bundle/examples/external_arguments.rs

```diff
@@ -20,23 +20,26 @@ unread-emails =
         .expect("Failed to add FTL resources to the bundle.");
 
     let mut args = HashMap::new();
-    args.insert("name", FluentValue::from("John"));
+    args.insert("name".to_string(), FluentValue::from("John"));
 
-    match bundle.format("hello-world", Some(&args)) {
-        Some((value, _)) => println!("{}", value),
-        _ => println!("None"),
-    }
+    let msg = bundle.get_message("hello-world").expect("Message exists");
+    let mut errors = vec![];
+    let pattern = msg.value.expect("Message has a value");
+    let value = bundle.format_pattern(&pattern, Some(&args), &mut errors);
```

Comment:
It would be nicer to use `if let Some(value) = msg.value { .. }` guard around the call to `format_pattern` in all examples, because it's much closer to a real-world usage of the API.

Reply:
I'm struggling with that a bit. I know this is a low level API that we expect to build on top of, but if someone would use it in their app, they wouldn't branch for a scenario where the string is missing, they'd expect the value of a string to be there if they placed it in their FTL file.

I don't want to suggest to localize an app by replacing:

```rust
println!("Enter your age:");
```

with:

```rust
if let Some(value) = msg.value {
  println!(bundle.format_pattern(value, None));
} else {
  unimplemented!()
}
```

Reply:
>  if someone would use it in their app, they wouldn't branch for a scenario where the string is missing, they'd expect the value of a string to be there if they placed it in their FTL file.

This only works if _their FTL file_ is the only FTL file in the app. As soon as they have more FTLs for more locales, this assumption may not hold anymore, depending on who wrote them. As you know, Fluent is designed around the idea that translations must not break the app. Handling that `Option<Pattern>` is an important enforcement of this design.

I know this is an _example_, and it's OK to simplify. But there is a chance people will look at examples to make their first steps around Fluent, and I'd like to take the opportunity to educate :) There's also a chance people will copy&paste directly from examples.

Also, it's approximately the same amount of code, and some might argue that the `if let` is actually nicer.

```rust
let pattern = msg.value.expect("Message has no value");
let value = bundle.format_pattern(&pattern, None);
```

```rust
if let Some(value) = msg.value {
  println!(bundle.format_pattern(value, None));
}
```

Or, to make the null-value case explicit, I'd put a comment in the `else` block, like so:

```rust
if let Some(value) = msg.value {
  println!(bundle.format_pattern(value, None));
} else {
  // Show the fallback translation, e.g. the id of the message.
}
```

Reply:
> As you know, Fluent is designed around the idea that translations must not break the app.

Yes, but we achieve it on a higher level. We introduce nice fallback mechanisms that handle the case where the user expects a value and there is none.

On this level there is no such mechanism, no such fallback.

I believe for a reader of an example, seeing:

```rust
if let Some(value) = msg.value {
  println!(bundle.format_pattern(value, None));
}
```

is bizarre. It's basically saying "try to localize the message, and if that fails, show nothing".
I believe that `.expect()` is exactly the right call for the example. It says "Since you placed the value, expect it to be there, but please, remember that it's an `Option` and it may be null, so you'll likely have a macro/higher-level-api that will facilitate a fallback".

```rust
if let Some(value) = msg.value {
  println!(bundle.format_pattern(value, None));
} else {
  // Show the fallback translation, e.g. the id of the message.
}
```

This is more acceptable, but still kind of explicit. It gives the vibe of "Fluent makes your code way less readable.".

I'd prefer to use `.expect` to indicate that we expect the message, and I intend to improve docs to clearly state that the purpose of FluentBundle is to be handled by a higher level abstraction that will handle fallbacking, and since fallbacking may differ between bindings and apps, we make FluentBundle API clean for the higher level API to pick their strategy.

Reply:
OK, that sounds fair, thanks!